packages/service-client/odsp-client/src/odspClient.ts

```diff
@@ -0,0 +1,205 @@
+	OdspResourceTokenFetchOptions,
+} from "@fluidframework/odsp-driver-definitions";
+import { ITokenResponse } from "@fluidframework/azure-client";
+// eslint-disable-next-line import/no-deprecated
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import { IRequest } from "@fluidframework/core-interfaces";
+import {
+	OdspClientProps,
+	OdspConnectionConfig,
+	OdspContainerServices,
+	OdspServiceAttributes,
+} from "./interfaces";
+import { OdspAudience } from "./odspAudience";
+
+/**
+ * OdspClient provides the ability to have a Fluid object backed by the ODSP service within the context of Microsoft 365 (M365) tenants.
+ *
+ * @alpha @sealed
+ */
+export class OdspClient {
+	private readonly documentServiceFactory: IDocumentServiceFactory;
+	private readonly urlResolver: OdspDriverUrlResolver;
+
+	public constructor(private readonly properties: OdspClientProps) {
+		const getSharePointToken = async (options: OdspResourceTokenFetchOptions) => {
+			const tokenResponse: ITokenResponse =
+				await this.properties.connection.tokenProvider.fetchStorageToken(
+					options.siteUrl,
+					"",
+				);
+			return {
+				token: tokenResponse.jwt,
+			};
+		};
+
+		const getPushServiceToken = async (options: OdspResourceTokenFetchOptions) => {
+			const tokenResponse: ITokenResponse =
+				await this.properties.connection.tokenProvider.fetchOrdererToken(options.siteUrl);
+			return {
+				token: tokenResponse.jwt,
+			};
+		};
+		this.documentServiceFactory = new OdspDocumentServiceFactory(
+			getSharePointToken,
+			getPushServiceToken,
+		);
+
+		this.urlResolver = new OdspDriverUrlResolver();
+	}
+
+	public async createContainer(containerSchema: ContainerSchema): Promise<{
+		container: IFluidContainer;
+		services: OdspContainerServices;
+	}> {
+		const loader = this.createLoader(containerSchema);
+
+		const container = await loader.createDetachedContainer({
+			package: "no-dynamic-package",
+			config: {},
+		});
+
+		const fluidContainer = await this.createFluidContainer(
+			container,
+			this.properties.connection,
+		);
+
+		const services = await this.getContainerServices(container);
+
+		return { container: fluidContainer, services };
+	}
+
+	public async getContainer(
+		sharingUrl: string,
+		containerSchema: ContainerSchema,
+	): Promise<{
+		container: IFluidContainer;
+		services: OdspContainerServices;
+	}> {
+		const loader = this.createLoader(containerSchema);
+		const container = await loader.resolve({ url: sharingUrl });
+
+		// eslint-disable-next-line import/no-deprecated
+		const rootDataObject = await requestFluidObject<IRootDataObject>(container, "/");
+		const fluidContainer = new FluidContainer(container, rootDataObject);
+		const services = await this.getContainerServices(container);
+		return { container: fluidContainer, services };
+	}
+
+	private createLoader(containerSchema: ContainerSchema): Loader {
+		const runtimeFactory = new DOProviderContainerRuntimeFactory(containerSchema);
```

Comment:
That's how we did it in AzureClient, but that looks wrong to me. Schema should be defined by file content, not by external code. It's not very clear what happens when they do not match.

Reply:
What do you mean "file content" defines schema?